### PR TITLE
expression: add builtin function `tidb_decode_key()`

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
* Add `tidb_decode_key()` function to decode key.

Related Parser PR: https://github.com/pingcap/parser/pull/560

### What is changed and how it works?
```sql
-- decode region key.
mysql>select region_id, tidb_decode_key(start_key) as k from `TIKV_REGION_STATUS` where tidb_decode_key(start_key) like 'tableID=%' limit 3;
+-----------+---------------------------------------------------------------+
| region_id | k                                                             |
+-----------+---------------------------------------------------------------+
| 107       | tableID=43, handle=300000000                                  |
| 128       | tableID=43, indexID=1, indexValues=0380000000165A0BC003000000 |
| 154       | tableID=43, indexID=1, indexValues=03800000000EE6B28003000000 |
+-----------+---------------------------------------------------------------+

mysql>select tidb_decode_key( '7480000000000000FF2E5F698000000000FF0000010380000000FF05F5E10003000000FF0000000000000000F7')
+----------------------------------------------------------------------------------------------------------------+
| tidb_decode_key( '7480000000000000FF2E5F698000000000FF0000010380000000FF05F5E10003000000FF0000000000000000F7') |
+----------------------------------------------------------------------------------------------------------------+
| tableID=46, indexID=1, indexValues=038000000005F5E10003000000                                                  |
+----------------------------------------------------------------------------------------------------------------+

mysql>select tidb_decode_key( '7480000000000000FF2E5F728000000011FFE1A3000000000000FA')
+----------------------------------------------------------------------------+
| tidb_decode_key( '7480000000000000FF2E5F728000000011FFE1A3000000000000FA') |
+----------------------------------------------------------------------------+
| tableID=46, handle=300000000                                               |
+----------------------------------------------------------------------------+

-- if the key is not record/index kye, just do nothing.
mysql>select tidb_decode_key( '7480000000000000FF2E5F698000000000FF0000010380000000FF05F5E10003000000FF0000000000000000F')
+---------------------------------------------------------------------------------------------------------------+
| tidb_decode_key( '7480000000000000FF2E5F698000000000FF0000010380000000FF05F5E10003000000FF0000000000000000F') |
+---------------------------------------------------------------------------------------------------------------+
| 7480000000000000FF2E5F698000000000FF0000010380000000FF05F5E10003000000FF0000000000000000F                     |
+---------------------------------------------------------------------------------------------------------------+

```

### Other discuss

The decoded key format currently maybe not very well, If you have any suggestions, Please comment, Thanks very much.

Maybe we can use the format  in `show table region.` 

```sql
+-----------+-------------------------------------+-------------------------------------+-----------+-----------------+---------------+------------+---------------+------------+----------------------+------------------+
| REGION_ID | START_KEY                           | END_KEY                             | LEADER_ID | LEADER_STORE_ID | PEERS         | SCATTERING | WRITTEN_BYTES | READ_BYTES | APPROXIMATE_SIZE(MB) | APPROXIMATE_KEYS |
+-----------+-------------------------------------+-------------------------------------+-----------+-----------------+---------------+------------+---------------+------------+----------------------+------------------+
| 95        | t_43_r                              | t_43_r_100000000                    | 112       | 9               | 98, 112, 111  | 0          | 129           | 0          | 1                    | 0                |
| 99        | t_43_r_100000000                    | t_43_r_200000000                    | 113       | 3               | 101, 113, 165 | 0          | 0             | 0          | 1                    | 0                |
| 103       | t_43_r_200000000                    | t_43_r_300000000                    | 114       | 9               | 105, 106, 114 | 0          | 126           | 0          | 1                    | 0                |
| 107       | t_43_r_300000000                    | t_43_r_400000000                    | 116       | 3               | 108, 116, 115 | 0          | 35            | 0          | 1                    | 0                |
| 2         | t_43_r_400000000                    |                                     | 73        | 8               | 73, 79, 162   | 0          | 0             | 0          | 1                    | 0                |
| 146       | t_43_i_1_                           | t_43_i_1_03800000000773594003000000 | 149       | 9               | 148, 149, 166 | 0          | 0             | 0          | 1                    | 0                |
| 150       | t_43_i_1_03800000000773594003000000 | t_43_i_1_03800000000ee6b28003000000 | 159       | 8               | 152, 153, 159 | 0          | 35            | 0          | 1                    | 0                |
| 154       | t_43_i_1_03800000000ee6b28003000000 | t_43_i_1_0380000000165a0bc003000000 | 160       | 1               | 155, 160, 161 | 0          | 35            | 0          | 1                    | 0                |
| 128       | t_43_i_1_0380000000165a0bc003000000 | t_43_i_2_                           | 145       | 3               | 131, 145, 144 | 0          | 0             | 0          | 1                    | 0                |
| 91        | t_43_i_2_                           | t_43_r                              | 93        | 8               | 92, 93, 94    | 0          | 35            | 0          | 1                    | 0                |
+-----------+-------------------------------------+-------------------------------------+-----------+-----------------+---------------+------------+---------------+------------+----------------------+------------------+
```

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

Release note
- Add `tidb_decode_key` build-in function to decode the key of TiDB used.

### CODE REVIEW

**Comment by crazycs520** [comment]
@winkyao PTAL

File: expression/builtin_info.go [hunk 1]
```diff
@@ -589,3 +596,61 @@ func (b *builtinRowCountSig) evalInt(_ chunk.Row) (res int64, isNull bool, err e
...
+	}
+	// Try decode as table index key.
+	tableID, indexID, indexValues, err := tablecodec.DecodeIndexKeyPrefix(key)
+	if err == nil {
+		idxValueStr := fmt.Sprintf("%X", indexValues)
+		return "tableID=" + strconv.FormatInt(tableID, 10) + ", indexID=" + strconv.FormatInt(indexID, 10) + ", indexValues=" + idxValueStr
+	}
+
+	// TODO: try to decode other type key.
+	return s
```
Comment (bb7133) [inline expression/builtin_info.go]: Please consider returning a warning message, for the reason described above.

**Comment by zz-jason** [comment]
@crazycs520 Please supplement the `release-note` section in the PR description. BTW, we also need to update the document about this new builtin function.

File: expression/builtin_info.go [hunk 2]
```diff
@@ -589,3 +596,63 @@ func (b *builtinRowCountSig) evalInt(_ chunk.Row) (res int64, isNull bool, err e
...
+	if isNull || err != nil {
+		return "", isNull, err
+	}
+	return decodeKey(b.ctx, s), false, nil
+}
+
+func decodeKey(ctx sessionctx.Context, s string) string {
+	key, err := hex.DecodeString(s)
+	if err != nil {
+		ctx.GetSessionVars().StmtCtx.AppendWarning(errors.New("the key is a invalid record/index key"))
```
Comment (zz-jason) [inline expression/builtin_info.go]:
```suggestion
		ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("invalid record/index key: %v", key))
```

File: expression/builtin_info.go [hunk 3]
```diff
@@ -589,3 +596,63 @@ func (b *builtinRowCountSig) evalInt(_ chunk.Row) (res int64, isNull bool, err e
...
+		return s
+	}
+	// Auto decode byte if needs.
```
Comment (zz-jason) [inline expression/builtin_info.go]:
```suggestion
	// Auto decode byte if needed.
```

File: expression/builtin_info.go [hunk 4]
```diff
@@ -589,3 +596,63 @@ func (b *builtinRowCountSig) evalInt(_ chunk.Row) (res int64, isNull bool, err e
...
+	_, bs, err := codec.DecodeBytes([]byte(key), nil)
+	if err == nil {
+		key = bs
+	}
+	// Try decode as table record key.
```
Comment (zz-jason) [inline expression/builtin_info.go]:
```suggestion
	// Try to decode it as a record key.
```

File: expression/builtin_info.go [hunk 5]
```diff
@@ -589,3 +596,63 @@ func (b *builtinRowCountSig) evalInt(_ chunk.Row) (res int64, isNull bool, err e
...
+	tableID, handle, err := tablecodec.DecodeRecordKey(key)
+	if err == nil {
+		return "tableID=" + strconv.FormatInt(tableID, 10) + ", handle=" + strconv.FormatInt(handle, 10)
```
Comment (zz-jason) [inline expression/builtin_info.go]: how about `s/handle/_tidb_rowid/`?

Comment (crazycs520) [reply expression/builtin_info.go]: If the primary key is the handle, there is no `tidb_rowid`  column.
So i think `handle` is better.

Comment (zz-jason) [reply expression/builtin_info.go]: From the user's perspective, we should provide a `_tidb_rowid` column if the PK is handle.

Comment (crazycs520) [reply expression/builtin_info.go]: OK, Done.

**Comment by crazycs520** [comment]
@zz-jason , `tidb_decode_region_key` is too narrow,  this function not only use to decode the region key, sometime in the TiDB log/TiKV log maybe record the conflict key, then we can use this function to decode too.

File: expression/builtin.go [hunk 6]
```diff
@@ -719,4 +719,7 @@ var funcs = map[string]functionClass{
 	ast.JSONDepth:         &jsonDepthFunctionClass{baseFunctionClass{ast.JSONDepth, 1, 1}},
 	ast.JSONKeys:          &jsonKeysFunctionClass{baseFunctionClass{ast.JSONKeys, 1, 2}},
 	ast.JSONLength:        &jsonLengthFunctionClass{baseFunctionClass{ast.JSONLength, 1, 2}},
+
+	// TiDB internal function.
+	ast.TiDBDecodeKey: &tidbDecodeKeyFunctionClass{baseFunctionClass{ast.TiDBDecodeKey, 1, 1}},
```
Comment (zz-jason) [inline expression/builtin.go]: There are other tidb only functions, can you move them together?

File: expression/builtin_info.go [hunk 7]
```diff
@@ -589,3 +596,63 @@ func (b *builtinRowCountSig) evalInt(_ chunk.Row) (res int64, isNull bool, err e
...
+	}
+	// Try decode as table index key.
+	tableID, indexID, indexValues, err := tablecodec.DecodeIndexKeyPrefix(key)
+	if err == nil {
+		idxValueStr := fmt.Sprintf("%X", indexValues)
+		return "tableID=" + strconv.FormatInt(tableID, 10) + ", indexID=" + strconv.FormatInt(indexID, 10) + ", indexValues=" + idxValueStr
+	}
+
+	// TODO: try to decode other type key.
+	ctx.GetSessionVars().StmtCtx.AppendWarning(errors.New("the key is a invalid record/index key"))
```
Comment (zz-jason) [inline expression/builtin_info.go]:
```suggestion
	ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("invalid record/index key: %v", key))
```

File: expression/builtin.go (hunk 6 above)
Comment (crazycs520) [reply expression/builtin.go]: done.

File: expression/builtin_info.go [hunk 8]
```diff
@@ -589,3 +596,63 @@ func (b *builtinRowCountSig) evalInt(_ chunk.Row) (res int64, isNull bool, err e
...
+	}
+	// Auto decode byte if needed.
+	_, bs, err := codec.DecodeBytes([]byte(key), nil)
+	if err == nil {
+		key = bs
+	}
+	// Try to decode it as a record key.
+	tableID, handle, err := tablecodec.DecodeRecordKey(key)
+	if err == nil {
+		return "tableID=" + strconv.FormatInt(tableID, 10) + ", _tidb_rowid=" + strconv.FormatInt(handle, 10)
```
Comment (winoros) [inline expression/builtin_info.go]: Why we don't return json object?

Comment (crazycs520) [reply expression/builtin_info.go]: Great idea.

Comment (crazycs520) [reply expression/builtin_info.go]: I will change this tomorrow.

**Pull request merged**